### DISCUSSION TITLE
How to constrain the rate of change of the control variable

### DISCUSSION THREAD

**Discussion by henryex2009** [casadi/casadi#3278] [Q&A]
Hello everyone, I want to control a power by MPC，the power is the control variable of  system.So how to constrain the rate of change of the power in CasaADi(matlab)，I have to try like this:
opti.subject_to(U(:,2)-U(:, 1) <= 0.25); 
but It seem dont work，could anybody  help to solve this problem.
Thank you

**Comment by jgillis** [comment]
`opti.subject_to(-0.25 <= U(:,2)-U(:, 1) <= 0.25);`?

**Reply by henryex2009** [reply]
hello,the point is, now solve the MPC,the result U is 0,5,5,5,5.......（simulate 250 step）,the fisrt step is 0 ,and the second is 5，what I  wish is the result U first is 0,and second is 0.25 ,thrid is 0.5.....every step the U only can add 0.25, not change to 5 directly.
whatever the -0.25 <= U(:,2)-U(:, 1) <= 0.25 or U(:,2)-U(:, 1) <= 0.25 seems not work.

**Reply by jgillis** [reply]
Please post a failing example

**Reply by henryex2009** [reply]
Hello,the example code is :
import casadi.*;
opti = casadi.Opti();
N=10; 
Ts=1;
P = opti.parameter(1, 1);
X0 = opti.parameter(1, 1);
U = opti.variable(1, N);
X = opti.variable(1, N + 1);
J = 0; 
Q1 = 10^5;
Q2=10^5;
u0 = zeros(1,10);
R = 1*eye(1);
for k=1:N
    opti.subject_to(X(:,k+1)==PTCnnFcn(X(:,k),U(:,k))); 
    J=J+U(:, k)' * R * U(:, k);
    J=J+(X(1, k + 1) - P(1))' * Q1 * (X(1, k + 1) - P(1))/30^2; 
end
opti.subject_to(X(:, 1) == X0);
opti.subject_to(-15 <= X(1, :) <= 22);
opti.subject_to(0 <= U(1, :) <= 5);
opti.subject_to(-0.25<=U(:,2)-U(:, 1) =< 0.25);
opti.subject_to(-0.25<=U(:,3)-U(:, 2) =< 0.25);
...
opti.minimize(J);
opti.set_initial(U, u0);
opts = struct;
opts.ipopt.max_iter = 100;
opts.ipopt.print_level =0;%0,3
opts.print_time = 0;
opts.ipopt.acceptable_tol =1e-8;
opts.ipopt.acceptable_obj_change_tol = 1e-6;
opti.solver('ipopt',opts);
t = 0;
x = -8.2; 
u0 = zeros(1,10);
reference = 20;
u = 0;
Ts_s= 1;

th=[t];
xh=[x];
uh=[u];

simtime=250;
    tic;
for i=1:(simtime/Ts_s)
    opti.set_value(P, reference);
    opti.set_value(X0, x);
    opti.set_initial(U, u0);
    sol = opti.solve();
    uout=sol.value(U);
    u=uout(:,1);
    x_update = PTCnnFcn(x,u);
    x=x_update;
    t=t+Ts;
    th=[th,t];
    xh=[xh,x];
    uh=[uh,u]; 
end
    toc;

**Reply by henryex2009** [reply]
Hello，I have added a new variable to solve this problem ,but I have a new question：how to generate the example to C code.
Accoding to the doc, I have to bulid a function ,so how to bulid the function with this example.
Thank you

**Reply by jgillis** [reply]
`PTCnnFcn` is missing to reproduce this failing example.

Please post new questions in a new thread.
Use `to_function` to be able to do codegen
